### PR TITLE
[torch] Add cuda support for segment reduction 'max'

### PR DESCRIPTION
**Pull request by serhaty**
Summary:
This is re submit of PR: https://github.com/pytorch/pytorch/pull/54175

Main changes compared to original PR:
- Switch to importing "<ATen/cuda/cub.cuh>"
- Use CUB_WRAPPER to reduce boiler plate code.

Test Plan:
Will check CI status to make sure a

Added unit test

Differential Revision: D27941257

### CODE REVIEW

**Comment by serhaty** [comment]
I am seeing two rocm related build issues but I am not sure what is the root cause since I don't see the output from UI. Is there a recommended way to debug such issues with rocm? Do I need to find an AMD capable device for this? cc: @jeffdaily for guidance on this.

**Comment by ngimel** [comment]
@zasdfgbnm can you please add a note on the proper way to include new cub functions now? This PR is building ok with cuda, but breaks with ROCM.

**Comment by zasdfgbnm** [comment]
I think for this PR, the problem is because `cub::DeviceSegmentedReduce` is not converted to `hipcub::DeviceSegmentedReduce` by hipify.

**Comment by zasdfgbnm** [comment]
I think you need to add your functions to https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py

**Comment by ngimel** [comment]
> I think you need to add your functions to https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py

And probably use NO_ROCM with CUB_WRAPPER

**Comment by jeffdaily** [comment]
CC @jithunnair-amd @sunway513

**Comment by serhaty** [comment]
> I think you need to add your functions to https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py

Just curious, would not it be better if default behavior was to do cub conversion automatically and only provide mapping if there is need for override (like API differences across the two)?

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -8937,4 +8937,4 @@
 - func: segment_reduce(Tensor data, str reduce, *, Tensor? lengths=None, Tensor? indices=None, int axis=0, bool unsafe=False) -> Tensor
```
Comment (ngimel) [inline aten/src/ATen/native/native_functions.yaml:8937]: as we've discussed, add `initial` argument here to handle 0-length case

**Review by ngimel — COMMENTED** [review]
As we've discussed, please add `initial` argument to the handle

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (serhaty) [reply aten/src/ATen/native/native_functions.yaml:8937]: will leave this change to another PR on the stack in favor of smaller PRs.

File: aten/src/ATen/native/cuda/SegmentReduce.cu [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#include <ATen/cuda/cub.cuh>
+#include <iostream>
+
+namespace at {
+namespace native {
+
+struct CustomMax {
+  template <typename OutputT>
+  __host__ __device__ __forceinline__ OutputT
+  operator()(const OutputT& a, const OutputT& b) {
```
[the remark this answers is not in the document]
Comment (serhaty) [reply aten/src/ATen/native/cuda/SegmentReduce.cu:17]: Thank you. Finally realized in logs that there is actually a way to see full logs  as well on jenkins.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**